### ISSUE TITLE
关于softmax

### ISSUE THREAD

**Issue by shiqingzhangCSU** [baichuan-inc/Baichuan-13B#140]

**Comment by KelleyYin** [comment]
mark

**Comment by NicholasYoungAI** [comment]
mark，同样遇到了精度问题

**Comment by shiqingzhangCSU** [comment]
> mark，同样遇到了精度问题

我这边解决了，通过用一个mask2 - max（mask2），这个操作是等价的，你可以试试。或者直接把softmax这部分计算全部还原为32。

**Comment by NicholasYoungAI** [comment]
> > mark，同样遇到了精度问题
> 
> 我这边解决了，通过用一个mask2 - max（mask2），这个操作是等价的，你可以试试。或者直接把softmax这部分计算全部还原为32。

我试过把softmax部分还原为fp32，仍然没有解决问题。

mask2 - max(mask2) 是指什么，我没有get到。

谢谢

**Comment by shiqingzhangCSU** [comment]
> > > mark，同样遇到了精度问题
> > 
> > 
> > 我这边解决了，通过用一个mask2 - max（mask2），这个操作是等价的，你可以试试。或者直接把softmax这部分计算全部还原为32。
> 
> 我试过把softmax部分还原为fp32，仍然没有解决问题。
> 
> mask2 - max(mask2) 是指什么，我没有get到。
> 
> 谢谢
1.softmax的相关的所有全部转为fp32试试。
2.就是你进行softmax的操作
attn_weights = torch.nn.functional.softmax(attn_weights, dim=-1)
这里的attn_weights减去自己的最大值max（attn_weights），这个操作对softmax是等价的，然后可以减少精度溢出。
